osquery/logger/plugins/buffered.cpp

```diff
@@ -67,14 +74,73 @@ void BufferedLogForwarder::check() {
       VLOG(1) << "Error sending status to logger: " << status.getMessage();
     } else {
       // Clear the status logs once they were sent.
-      iterate(indexes, ([&results, this](std::string& index) {
+      iterate(indexes, ([this](std::string& index) {
                 if (!isStatusIndex(index)) {
                   return;
                 }
-                deleteDatabaseValue(kLogs, index);
+                std::lock_guard<std::recursive_mutex> lock(count_mutex_);
+                deleteValueWithCount(kLogs, index);
               }));
     }
   }
+
+  // Purge any logs exceeding the max after our send attempt
+  if (FLAGS_buffered_log_max > 0) {
+    purge();
+  }
+}
+
+void BufferedLogForwarder::purge() {
+  std::lock_guard<std::recursive_mutex> lock(count_mutex_);
+  size_t buffer_count = getBufferCount();
+  if (buffer_count <= FLAGS_buffered_log_max) {
+    return;
+  }
+
+  size_t purge_count = buffer_count - FLAGS_buffered_log_max;
+
+  // Collect purge_count indexes of each type (result/status) before
+  // partitioning to find the oldest. Note this assumes that the indexes are
+  // returned in ascending lexicographic order (true for RocksDB).
+  std::vector<std::string> indexes;
+  auto status =
+      scanDatabaseKeys(kLogs, indexes, genIndexPrefix(true), purge_count);
+  if (!status.ok()) {
+    LOG(ERROR) << "Error scanning DB during buffered log purge";
+    return;
+  }
+
+  std::vector<std::string> status_indexes;
+  status = scanDatabaseKeys(kLogs, status_indexes, genIndexPrefix(false),
+                            purge_count);
+  if (!status.ok()) {
+    LOG(ERROR) << "Error scanning DB during buffered log purge";
+    return;
+  }
+
+  indexes.insert(indexes.end(), status_indexes.begin(), status_indexes.end());
+
+  size_t prefix_size = genIndexPrefix(true).size();
+  // Partition the indexes so that the first purge_count elements are the
+  // oldest indexes (the ones to be purged)
+  std::nth_element(indexes.begin(), indexes.begin() + purge_count - 1,
+                   indexes.end(),
+                   [&](const std::string& a, const std::string& b) {
+                     // Skip the prefix when doing comparisons
+                     return a.compare(prefix_size, std::string::npos, b,
+                                      prefix_size, std::string::npos) < 0;
+                   });
+  indexes.erase(indexes.begin() + purge_count, indexes.end());
+
+  // Now only indexes of logs to be deleted remain
+  iterate(indexes, [this](const std::string& index) {
+    if (!deleteValueWithCount(kLogs, index).ok()) {
+      LOG(ERROR) << "Error deleting value during buffered log purge";
+    }
+  });
+
+  LOG(WARNING) << purge_count << " buffered logs purged (max "
```

Comment:
Did you see my note from before?

> How about:
> ```cpp
> LOG(WARNING) << "Purging buffered logs limit (" << FLAGS_buffered_log_max << ") exceeded: " << indexes.size();
> ```
> It's 'slightly' weird but matches with some of our other warning logs. :/

My real concern is having non-`INFO`/`VLOG` message begin with a word rather than number.